README.md
# CreateGeneSignatures

![CI](https://github.com/ATpoint/CreateGeneSignatures/actions/workflows/ci.yml/badge.svg)

## Introduction

This package creates per-group marker signatures from pairwise differential expression (DE) results. For this it expects a list of DE result tables (data.frames) for all possible pairwise combinations of groups. For example, for three groups A, B and C that would be A-B, A-C and B-C. It first subsets these DE results to "significant" genes based on user-defined cutoffs, for example based on FDR and logFC. It then ranks these DE results by a user-defined metric, such as `-log10(pvalue)` and then aggregates the results into ranked list of markers per group, see `?RankDEGs`. Based on these lists the `CreateGeneSignatures()` function then filters for genes that are DE in a given groups versus a proportion (`min.prop`) of other groups, for example 1 means all groups, and 0.75 means 75% of other groups. The ranking of the per-group markers is based on the chosen ranking metric. Gene rank highly if they consistently ranked highly in the individual DE tables.

The aim of this package is simplicity and understandable "non-blackbox" behaviour. Therefore, the marker criteria are simply "a gene must be DE with cutoffs (...) and for a given group this must be true versus (all, half, 3/4, min.prop) of other groups.

## Example workflow:

The example workflow uses RNA-seq data from Haemopedia. It first perform differential analysis with `edgeR`,
then filters the DE results for significant genes, ranks the DEGs and then create a signature for every of the four celltypes, namely CD4-T cells, CD8-T cells, naive B cells and natural killer (NK) cells.

```{r}
# load RNA-seq data for CD4T-, CD8T and naive B cells from Haemopedia:
counts <- readRDS(paste0(
  system.file("extdata",package="CreateGeneSignatures"),
  "/haemopedia_subset.rds"))

# Use edgeR to perform all pairwise comparisons:
library(edgeR)

y <- DGEList(counts=counts,group=gsub("\\..", "", colnames(counts)))
design <- model.matrix(~0+group,y$samples)
colnames(design) <- gsub("group", "", colnames(design))
y <- y[filterByExpr(y),,keep.lib.size=FALSE]
y <- calcNormFactors(y)
y <- estimateDisp(y,design)
fit <- glmQLFit(y,design)

# all unique pairwise contrasts:
contrasts <- makeContrasts(CD4T_vs_CD8T  = CD4T-CD8T,
                           CD4T_vs_NK    = CD4T-NK,
                           CD4T_vs_NveB  = CD4T-NveB,
                           CD8T_vs_NK    = CD8T-NK,
                           CD8T_vs_NveB  = CD8T-NveB,
                           NK_vs_NveB    = NK-NveB,
                           levels = design)

# Get DEGs, testing against a fold change:                  
res <- sapply(colnames(contrasts), function(con){
  tt<-topTags(glmTreat(fit,contrast=contrasts[,con], log2(1.5)),n=Inf)$table
  return(data.frame(Gene=rownames(tt), tt))
}, simplify = FALSE)

# Rank the DEGs:
ranked <- RankDEGs(res, delim="_vs_", signif.column="FDR", signif.threshold=0.05,
                   effect.column="logFC", effect.threshold=0, gene.column="Gene",
                   rnk.column="PValue", rnk.method="increasing")

# Create signatures, keeping top 50 signature genes that separate the respective celltype from all other celltypes:
signatures <- CreateGeneSignatures(ranked=ranked, keep.n=50, min.prop=1)

# check number of genes. for CD8T cells we found < 50 genes:
lengths(signatures)

# Inspect signatures using heatmaps plotting the scaled logcpms of the signature genes:
library(pheatmap)
logcpm <- log2(edgeR::cpm(y,log=FALSE)+1)

# plot a heatmap ordered by groups:
col_order <- unlist(lapply(names(ranked), function(x) grep(paste0("^", x), colnames(logcpm))))

# use scaled logCPMs:                     
logcpmZ <- t(scale(t(logcpm[unique(unlist(signatures)),])))
pheatmap(mat=logcpmZ[,col_order],
         show_rownames=FALSE, cluster_rows=FALSE, cluster_cols=FALSE)     
```

A heatmap of the combined signatures genes:

![heatmap](misc/heatmap.png)

In the above example the parameters were most strict, with `min.prop=1` requiring that signature genes were DE in all comparisons. Alternatively, if this returns no, or too few genes one might relax this cutoff, for example `min.prop=0.75`, which requires that a gene is DE in at least 75% of comparisons. 

## Installation

```r
install.packages("remotes")
remotes::install_github("ATpoint/CreateGeneSignatures")
```


NEWS
Log of changes between versions:

v2.2.3:
- minor changes in docs and change default of `CreateGeneSignatures` argument `extended` to `FALSE`.

v2.2.2:
- minor changes in DESCRIPTION and render plots via GitHub Actions
- extend test.R script to render the plot in README.md and update via Actions

v2.2.1:
- fixing bugs and edge cases, robustified

v2.2.0:
- added `extended` argument to output a full data.frame that indicates whether a gene qualified as a marker for the particular
group against the other groups. Makes sense if min.prop < 1 to see which groups exactly the gene does not separate. Also some 
fixes to use_groups, now if min.prop < 1 a gene must have the same "off-target" pattern to qualify as a marker for these combined groups.
Some updates in the examples.

v2.1.0: 
- added exclude_groups to CreateGeneSignatures() to remove groups from the signature process

v2.0.1:
-fix typos in example section

v2.0.0:
- now support defining signatures for a group of celltypes, e.g. in the example set CD4T and CD8T versus the rest.
This has the advantage that these two groups are not compared against each other which is beneficial as they share a lot of markers
and one only cares about genes that separate both from the rest but not from each other.

v1.1.0:
-made the ranking function robust against syntactically-invalid R names which in the previous version would have returned no genes at all

v1.0.1: 
-changed code example for plotting a heatmap based on the signatures

v1.0.0:
-initial Github commit


R/tests.R
args <- commandArgs(trailingOnly=TRUE)
run_test <- if(is.na(args[1])) FALSE else as.logical(args[1])

if(run_test){
  
  if(sum(c("edgeR", "pheatmap") %in% rownames(installed.packages())) < 2){
    install.packages("BiocManager", repos="https://cloud.r-project.org/")
    BiocManager::install(c("edgeR", "pheatmap"))
  }
    
  library(edgeR)
  library(pheatmap)

  source("R/RankDEGs.R")
  source("R/CreateGeneSignatures.R")
  
  res <- readRDS("inst/extdata/res.rds")
  
  #/ rank DEGs
  ranked <- RankDEGs(res)
  
  #/ test1: standard use
  signatures <- CreateGeneSignatures(ranked=ranked, keep.n=50, min.prop=1, extended=FALSE)
  
  #/ test2: pool groups
  signatures2 <- CreateGeneSignatures(ranked=ranked, use_groups=c("CD4T", "CD8T"), extended=FALSE)
  
  #/ test3: exclude groups
  signatures3 <- CreateGeneSignatures(ranked=ranked, exclude_groups=c("CD4T"), extended=FALSE)
  
  #/ test 4: with extended
  signatures4 <- CreateGeneSignatures(ranked=ranked, min.prop=2/3, extended=TRUE)

  counts <- readRDS("inst/extdata/haemopedia_subset.rds")
  
  y <- DGEList(counts=counts,group=gsub("\\..", "", colnames(counts)))
  
  design <- model.matrix(~0+group,y$samples)
  
  colnames(design) <- gsub("group", "", colnames(design))
  
  y <- y[filterByExpr(y),,keep.lib.size=FALSE]
  
  y <- calcNormFactors(y)
  
  y <- estimateDisp(y,design)
  
  fit <- glmQLFit(y,design)
  
  contrasts <- makeContrasts(CD4T_vs_CD8T  = CD4T-CD8T,
                             CD4T_vs_NK    = CD4T-NK,
                             CD4T_vs_NveB  = CD4T-NveB,
                             CD8T_vs_NK    = CD8T-NK,
                             CD8T_vs_NveB  = CD8T-NveB,
                             NK_vs_NveB    = NK-NveB,
                             levels = design)
  
  res <- sapply(colnames(contrasts), function(con){
    tt<-topTags(glmTreat(fit,contrast=contrasts[,con],lfc=log2(1.5)),n=Inf)$table
    return(data.frame(Gene=rownames(tt), tt)[,c("Gene", "logFC", "PValue", "FDR")])
  }, simplify = FALSE)
  
  logcpm <- log2(edgeR::cpm(y,log=FALSE)+1)
  
  col_order <- unlist(lapply(names(ranked), 
                             function(x) grep(paste0("^", x), colnames(logcpm))))
  
  logcpmZ <- t(scale(t(logcpm[unique(unlist(signatures)),])))
  
  p <- pheatmap(mat=logcpmZ[,col_order],
                show_rownames=FALSE, cluster_rows=FALSE, cluster_cols=FALSE)
  
  png("heatmap.png")
  print(p); dev.off()

}


R/RankDEGs.R
#' Rank genes based on differential expression statistics
#' 
#' Rank genes from lists of pairwise comparison by significance or effect size
#' 
#' @param res a named list of pairwise DE results, see details.
#' @param delim a string that delimits the comparison groups in \code{names(res)}, e.g. celltype1_vs_celltype2 would be "_vs_"
#' @param signif.column colname storing significances to use for filtering, e.g. FDR
#' @param signif.threshold keep only genes with \code{signif.column} below that threshold
#' @param effect.column colname storing the effect size, e.g. logFC. Must be a zero-centered effect size so effect size > 0 means higher in one group and < 0 means lower.
#' Don't use something like AUCs from a Wilcox test where > 0.5 means higher and < 0.5 means lower per group.
#' @param effect.threshold keep only genes with effect.column above this threshold, could be a minimum effect size even though it is recommended to explicitely test against
#' the desired minimum effect size rather than postfiltering, see details.
#' @param gene.column colname storing genes or any kind of row identifiers, those will be returned in the output meeting the above criteria
#' @param rnk.column use this column for the ranking
#' @param rnk.method either 'decreasing' or 'increasing' ranking based on \code{rnk.column}
#' 
#' @details 
#' For an example of how the input should look like see the examples. The pairwise comparisons must be unique, so if something like
#' celltype1_vs_celltype2 is present then do not include celltype2_vs_celltype1 into \code{res} as this is identical and only the sign of the 
#' effect size (e.g. the logFC) changes. The function handles this internally for every celltype.
#' 
#' The \code{signif.column} and \code{effect.column} are used first to filter the data, e.g. for FDR and logFC, and then the ranking is done based
#' on the \code{rnk.column} but aware of the direction of change based on \code{effect.column}, and in its current state only genes
#' with a positive effect size are taken into account. The \code{rnk.column} could e.g. e the nominal PValue or t-stat column which both
#' have the advantage over FDR that they usually have no ties.
#'
#' The output will be a nested list with the ranked genes for every celltype compared to every other celltype based in the entries of \code{res},
#' see the examples. Something like:
#' $celltype1
#' ..$celltype2
#' ..$celltype3
#' $celltype2
#' ..$celltype1
#' ..$celltype3
#' $celltype3
#' ..celltype1
#' ..celltype2
#'
#' For an example with real data see the examples of the \code{CreateGeneSignatures} function of this package.#' 
#' @author Alexander Toenges
#' 
#' @examples 
#' # first make some dummy DE results, then rank:
#' set.seed(1)
#' res <- sapply(c("gr1_vs_gr2","gr2_vs_gr3","gr1_vs_gr3"), function(x){
#'   data.frame(Gene=paste0("Gene",1:10), 
#'              logFC=rnorm(10,1,2),
#'              PValue=jitter(rep(0.001, 10), 20),
#'              FDR=jitter(rep(0.04, 10), 20))
#' },simplify=FALSE)
#' 
#' # this is how the results tables look:
#' res$gr1_vs_gr2
#' 
#' ranked <- RankDEGs(res=res, rnk.column="PValue", rnk.method="increasing")
#' 
#' @export
RankDEGs <- function(res, delim="_vs_", 
                     signif.column="FDR", signif.threshold=0.05,
                     effect.column="logFC", effect.threshold=0,
                     gene.column="Gene", rnk.column="PValue", rnk.method="increasing"){
  
  #---------------------------
  # Checks
  #---------------------------
  
  if(!class(res) %in% c("list", "SimpleList") | is.null(names(res))){
    stop("res must be a named list", call.=FALSE)
  }
  
  if(!all(grepl(delim, names(res)))) stop("delim was not found in all names of the res")
  
  invisible(match.arg(arg=class(signif.threshold), choices="numeric"))
  invisible(match.arg(arg=class(effect.threshold), choices="numeric"))
  
  check.gf <- sum(unlist(lapply(res, function(x) 
    if(!gene.column %in% colnames(x)) return(1) else return(0))))
  check.sf <- sum(unlist(lapply(res, function(x) 
    if(!signif.column %in% colnames(x)) return(1) else return(0))))
  check.ef <- sum(unlist(lapply(res, function(x) 
    if(!effect.column %in% colnames(x)) return(1) else return(0))))
  check.rk <- sum(unlist(lapply(res, function(x) 
    if(!rnk.column %in% colnames(x)) return(1) else return(0))))
  
  if(check.gf>0) stop("gene.column does not exist in all entries of res.list!")
  if(check.sf>0) stop("signif.column does not exist in all entries of res.list!")
  if(check.ef>0) stop("effect.column does not exist in all entries of res.list!")
  if(check.rk>0) stop("rnk.column does not exist in all entries of res.list!")
  
  if(!rnk.method %in% c("decreasing", "increasing"))
    stop("rnk.method must be one of <increasing, decreasing>")
  
  #---------------------------
  # Ranking
  #---------------------------
  
  #/ make robust against syntactically non-valid names
  unq <- sort(unique(unlist(strsplit(names(res), delim))))
  lookup <- data.frame(original=unq, new=make.names(unq))
  
  if(!all(lookup$original==lookup$new))
    warning(paste("Detected syntactically invalid names.",
                  "It will work anyway, but be sure to double-check your results!",
                  "The invalid names are:",
                  paste0(lookup$original[which(!lookup$original==lookup$new)], collapse="\n"),
                  sep="\n"))
  
  unq <- make.names(unq)
  names(res) <- make.names(names(res))
  
  l <- list()
  for(i in unq){
    
    nm <- grep(paste(paste0("^",i,delim),paste0(delim,i,"$"), sep="|"), 
               names(res), value=TRUE)
    
    l[[i]] <- 
      lapply(X=nm, FUN=function(x){
        
        s  <- strsplit(x,delim)[[1]]
        
        tt <- res[[x]]
        
        if(i==s[2]) tt[,effect.column] <- tt[,effect.column]*-1
        
        if(!is.null(signif.column)) {
          sig <- (tt[,signif.column] < signif.threshold)
        } else sig <- rep(TRUE, nrow(tt))
        
        if(!is.null(effect.column)) {
          eff <- (tt[,effect.column] > effect.threshold)
        } else eff <- rep(TRUE, nrow(tt))
        
        tt<-tt[rowSums(data.frame(sig,eff))==2,]
        
        if(nrow(tt)>0){
          
          decreasing <- if(rnk.method=="decreasing") TRUE else FALSE
          
          ix <- sort(tt[,rnk.column], index.return=TRUE)$ix
          
          ranked <- tt[ix,gene.column]
          
        } else ranked <- NULL
        
        return(ranked)
        
      })
    
    names(l[[i]]) <- setdiff(unlist(strsplit(nm,delim)), i)
    
  }
  
  iszero <- sum(!lengths(l) > 0)
  if(iszero>0) message("There are ", iszero, " comparisons with zero genes")
  
  #/ map names back to original
  names(l) <- lookup$original[match(names(l), lookup$new)]
  sapply(l, function(x)  {
    
    names(x) <- lookup$original[match(names(x), lookup$new)]
    x
    
  }, simplify=FALSE)
  
}

R/CreateGeneSignatures.R
#' Create gene signatures in a rank-based style
#' 
#' Convert lists of ranked genes into celltype-specific signatures
#' 
#' @param ranked nested lists of ranked genes for each celltype, see details and the RankDEGs function
#' @param use_groups trigger group-aware mode. A vector with celltype names to be used to define markers for as a group,
#' so this group versus all others. See details.
#' @param exclude_groups vector, remove these groups from the signature creation
#' @param delim a string indicating the delimiter of names(ranked), e.g. celltype1_vs_celltype2 would be "_vs_"
#' @param keep.n number of genes to keep per signature, by default all candidates are returned for manual posthoc filtering
#' @param min.prop minimum proportion of comparisons per celltype that a gene must be included in. See details.
#' @param extended logical, whether to output the signature genes as a data.frame that indicates whether the gene
#' qualified as a marker against the other groups (1) or not (0)
#' 
#' @author Alexander Toenges
#' 
#' @examples 
#' # load RNA-seq data for CD4T-, CD8T and naive B cells from Haemopedia:
#' counts <- readRDS(paste0(
#'             system.file("extdata",package="CreateGeneSignatures"),
#'             "/haemopedia_subset.rds"))
#'
#' # Use edgeR to perform all pairwise comparisons
#' 
#' library(edgeR)
#' y <- DGEList(counts=counts,group=gsub("\\..", "", colnames(counts)))
#' design <- model.matrix(~0+group,y$samples)
#' colnames(design) <- gsub("group", "", colnames(design))
#' y <- y[filterByExpr(y),,keep.lib.size=FALSE]
#' y <- calcNormFactors(y)
#' y <- estimateDisp(y,design)
#' fit <- glmQLFit(y,design)
#' 
#' # all unique pairwise contrasts:
#' contrasts <- makeContrasts(CD4T_vs_CD8T  = CD4T-CD8T,
#'                            CD4T_vs_NK    = CD4T-NK,
#'                            CD4T_vs_NveB  = CD4T-NveB,
#'                            CD8T_vs_NK    = CD8T-NK,
#'                            CD8T_vs_NveB  = CD8T-NveB,
#'                            NK_vs_NveB    = NK-NveB,
#'                            levels = design)
#'                            
#' # test using glmTreat against a minumum fold change of ~1.5                          
#' res <- sapply(colnames(contrasts), function(con){
#'   tt<-topTags(glmTreat(fit,contrast=contrasts[,con],lfc=log2(1.5)),n=Inf)$table
#' return(data.frame(Gene=rownames(tt), tt))
#' }, simplify = FALSE)
#' 
#' # Rank the DEGs:
#' ranked <- RankDEGs(res)
#' 
#' # Create signatures, keeping top 50 signature genes that separate the respective celltype
#' # from all other celltypes:
#' signatures <- CreateGeneSignatures(ranked=ranked, keep.n=50)
#' # check number of genes. for CD8T cells we found < 50 genes:
#' lengths(signatures)
#' 
#' # Inspect signatures using heatmaps plotting the scaled logcpms of the signature genes
#' library(pheatmap)
#' logcpm <- log2(edgeR::cpm(y,log=FALSE)+1)
#' 
#' # plot a heatmap in the order of names(signatures)
#' col_order <- unlist(lapply(names(ranked), 
#'                            function(x) grep(paste0("^", x), colnames(logcpm))))
#'                            
#' # use scaled logCPMs                           
#' logcpmZ <- t(scale(t(logcpm[unique(unlist(signatures)),])))
#' pheatmap(mat=logcpmZ[,col_order],
#'          show_rownames=FALSE, cluster_rows=FALSE, cluster_cols=FALSE)
#'          
#' # or each signature individually:
#' pheatmap(mat=t(scale(t(logcpm[signatures$CD4T,]))), show_rownames=FALSE)
#' pheatmap(mat=t(scale(t(logcpm[signatures$CD8T,]))), show_rownames=FALSE)
#' pheatmap(mat=t(scale(t(logcpm[signatures$NveB,]))), show_rownames=FALSE)
#' pheatmap(mat=t(scale(t(logcpm[signatures$NK,]))), show_rownames=FALSE)
#'          
#' # or genes that separate the CD4T and CD8T cells from the rest
#' signatures2 <- CreateGeneSignatures(ranked=ranked, use_groups=c("CD4T", "CD8T"), extended=FALSE)
#' logcpmZ2 <- t(scale(t(logcpm[signatures2,])))
#' pheatmap(mat=logcpmZ2[,col_order],
#'          show_rownames=TRUE, cluster_rows=FALSE, cluster_cols=FALSE)
#'          
#' # find signatures for each group but exclude the CD4T categorically
#' signatures3 <- CreateGeneSignatures(ranked=ranked, exclude_groups=c("CD4T"), extended=FALSE)
#' 
#' # Find all markers for each group against all but one group and output the extended table
#' # to easily see which group the gene does not qualify as a marker for.
#' # Here, the genes in the below tail command are markers for CD4T against NK and NveB but do not separate CD4T from CD8T.
#' signatures4 <- CreateGeneSignatures(ranked=ranked, min.prop=2/3, extended=TRUE)
#' tail(signatures4$CD4T)
#' 
#' @details 
#' The function takes as input a nested list that, for every celltype, contains the ranked genes from pairwise comparisons against every other
#' celltype in this analysis, see also the \code{RankDEGs} function of this package. For every celltype first a rank matrix is created that
#' stores the ranked genes from every pairwise comparison, assigning low scores to genes that ranked high in the input ranking and vice versa.
#' Genes to form the signatures are then selected by choosing those genes with minimal ranks across all comparisons.
#' In an iterative fashion genes that ranked high all comparisons are chosen first (and then ordered by median score), then genes ranked high in all but one, 
#' in all but two, all but three etc comparisons. The user can choose via \code{min.prop} the fraction of comparisons that a gene be highly ranked in.
#' For example a setup with four celltypes and \code{min.prop=1} would mean that only genes are included that ranked high in all comparisons.
#' If \code{min.prop=2/3} is chosen then signature genes must rank high in at least two out of three comparisons. The higher this value between (between zero and one)
#' the more stringent the filtering. If celltypes are decently separated it might be trivial to find large numbers of signature genes even with min.prop of 1.
#' In cases where cells are very similar to each other, e.g. cells being close in a developmental continuum, one will need to lower this threshold in order to get some
#' (or any) signature genes. 
#' 
#' The \code{use_groups} argument allows to specify a group of celltypes and the function will then infer 
#' markers for this combined group versus the rest. In the example above we use it to define combined CD4/CD8 markers. The advantage here is that these
#' cells have many markers in common so by defining them as a group we only look for "pan"-T cell markers and that retains more genes
#' that when treating both celltypes as independent groups. 
#' 
#' @export
CreateGeneSignatures <- function(ranked, 
                                 use_groups=NULL, exclude_groups=NULL,
                                 delim="_vs_", keep.n=Inf, min.prop=1,
                                 extended=FALSE){
  
  #/ checks
  if(!is.null(use_groups)){
    if(!sum(use_groups %in% names(ranked))==length(use_groups))
      stop("Make sure use_groups are part of names(ranked)")
  }
  
  if(!is.null(exclude_groups)){
    if(!sum(exclude_groups %in% names(ranked))==length(exclude_groups))
      stop("Make sure exclude_groups are part of names(ranked)")
  }
  
  if(!is.null(use_groups) & !is.null(exclude_groups)){
    if(length(intersect(use_groups, exclude_groups))>1)
      stop("There are overlaps between elements in use_groups and exclude_groups!")
  }
  
  if(is.null(names(ranked))) stop("ranked has no names")
  if(min.prop > 1 | min.prop < 0) stop("min.prop must be between > 0 and <= 1")
  
  #/ optionally remove exclude_groups from ranked
  if(!is.null(exclude_groups)){
    
    keep_them <- setdiff(names(ranked), exclude_groups)
    
    ranked <- 
      sapply(keep_them, function(x){
        
        r <- ranked[[x]]
        r[setdiff(keep_them, x)]
        
      }, simplify=FALSE)
    
  }
  
  #/ optionally only look at the group(s) in use_groups rather than all vs all
  if(is.null(use_groups)){
    
    nms <- names(ranked)
    
  } else nms <- use_groups
  
  #/ find the signature genes
  s <- sapply(nms, function(x){
    
    genes.ranged <- ranked[[x]]
    
    remove_these <- if(is.null(use_groups)) x else nms
    
    genes.ranged[remove_these] <- NULL # if the list contains the reference cell type itself
    
    #/ rank matrix, low values means high ranks and viceversa, 1 means not present in that pairwise list: 
    rmat <- rankmatrix(genes.ranged=genes.ranged)
    
    # Get the genes that qualify as markers given the number of groups and the min.prop
    isna <- !is.na(rmat)
    from <- ncol(rmat)
    to   <- floor(min.prop*from)
    to <- if(to==0) 1 else to
    
    med  <- function(y) median(y,na.rm=TRUE)
    
    #/ take the rankmatrix (so each entry is the rank of the gene in the individual ranking),
    #/ and then calculate the median of the ranks -- that is the final ranking metric
    final <- lapply(from:to, function(n){
      
      idx <- base::rowSums(isna)==n
      if(sum(idx)==0) return(NULL)
      
      rmat_x <- rmat[idx,,drop=FALSE]
      
      #/ get the median of the ranks
      mat <- apply(rmat_x, 1, med)
      
      #/ rank by the median
      o <- order(mat)
      
      markers.df <- data.frame(rmat_x[o,,drop=FALSE], check.names=FALSE)
      
      return(markers.df)
      
    })
    
    final <- do.call(rbind, final)
    
    final[!is.na(final)] <- 1
    final[is.na(final)]  <- 0
    final <- head(final, n=keep.n)
    
    return(final)
    
  },simplify = FALSE)
  
  #/ if use_groups is set then only return genes that qualify as markers in all members of 
  #/ use_groups with the same "off-target" pattern so if min.prop demands gene being a marker
  #/ against all but one group then this "one" group must be the same for all elements of use_groups
  if(!is.null(use_groups) & length(use_groups)>1 & min.prop < 1){
    
    #/ strict intersect
    isec <- Reduce(intersect, lapply(as.list(s), function(x) rownames(x)))
    
    if(length(isec)==0) {
      message("No signature genes found!")
      return(NULL)
    }
    
    #/ get the genes
    lp <- sapply(s, function(x) x[isec,], simplify=FALSE)
    cb <- combn(names(s), 2)
    
    lpx_check <- 
      lapply(1:ncol(cb), function(x) {
        
        y <- cb[,x]
        lpx <- lp[[y[1]]]==lp[[y[2]]]
        
        rownames(lpx[rowSums(lpx)==ncol(lpx),])
        
      })
    
    isec_x <- Reduce(intersect, lpx_check)
    
    to_return <- s[[1]][isec_x,]
    
  } else to_return <- s
  
  if(!is.null(use_groups) & min.prop==1 & length(use_groups)>0){
    
    j <- Reduce(intersect, lapply(to_return, rownames))
    to_return <- to_return[[1]][j,]
    
  }
  
  #/ either return the full table or just the names
  if(!extended){
    
    if(class(to_return)[1]=="list"){
      to_return <- sapply(to_return, function(x) rownames(x), simplify=FALSE)  
    } else to_return <- rownames(to_return)
    
    
  }
  
  to_return <- if(length(to_return)==1) to_return[[1]] else to_return
  
  return(to_return)
  
}

#' For the genes being present in genes.ranged per celltype we rank from 1 (best rank) to length 
#' of the current genes.ranged entry, and NA for genes not in that respective entry but in u.
#' Modified from Kolde et al in RobustRankAggreg::rankMatrix()
rankmatrix <- function (genes.ranged) 
{
  u = unique(c(genes.ranged, recursive = TRUE))
  N = length(u)
  
  #/ matrix with only NAs, to be filled with the ranks:
  rmat = matrix(NA, nrow=length(u), ncol=length(genes.ranged), 
                dimnames=list(u, names(genes.ranged)))
  
  #/ fill matrix with ranks per gene and celltype.
  #/ If gene is missing in a celltype (=was not signif vs other celltypes),
  #/ then leave at NA:
  for (i in names(genes.ranged)) {
    
    rmat[genes.ranged[[i]], i] = seq(1, length(genes.ranged[[i]]))
    
  }
  return(rmat)
}

